File: Programmers/42586.py
from collections import deque

def solution(progresses, speeds):
    answer = []

    time = 0
    queue = deque(progresses)
    s = deque(speeds)
    cnt = 0
    while queue:
        time += 1
        while queue and queue[0] + s[0]*time >= 100:
            cnt += 1
            queue.popleft()
            s.popleft()
        if cnt:
            answer.append(cnt)
            cnt = 0
    return answer

File: Programmers/42586.java
import java.util.*;

class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        List<Integer> answer = new ArrayList<>();

        Deque<Integer> queue = new ArrayDeque<>();
        Deque<Integer> s = new ArrayDeque<>();
        for (int p : progresses) queue.addLast(p);
        for (int sp : speeds) s.addLast(sp);

        int time = 0;
        int cnt = 0;
        while (!queue.isEmpty()) {
            time++;
            while (!queue.isEmpty() && queue.peekFirst() + s.peekFirst() * time >= 100) {
                cnt++;
                queue.pollFirst();
                s.pollFirst();
            }
            if (cnt > 0) {
                answer.add(cnt);
                cnt = 0;
            }
        }

        return answer.stream().mapToInt(Integer::intValue).toArray();
    }
}
